cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageInfo.java

```diff
@@ -313,55 +317,63 @@ private long computeChecksum() {
+        Set<Operation> incomingOperations = new LinkedHashSet<>();
+        incomingOperations.add(write);
+        computeIncoming(new EndPointField(write.getDestination(), input.getName()),
+                        operationsMap.get(input.getOrigin()), write, incomingOperations, result);
       }
     }
     return result;
   }
 
-  private void computeIncomingSummaryHelper(EndPointField destination, Operation currentOperation,
-                                            Operation previousOperation, Set<String> visitedOperationNames,
-                                            Map<EndPointField, Set<EndPointField>> result) {
-    if (!visitedOperationNames.add(currentOperation.getName())) {
+  /**
+   * Helper method to compute the incoming connections
+   *
+   * @param destination the {@link EndPointField} whose incoming connection needs to be processed
+   * @param currentOperation the operation being processed (since we are processing incoming this operation is on the
+   * left side if graph is imagined in horizontal orientation or this operation is the input to the to
+   * previousOperation
+   * @param previousOperation the previous operation which is processed and reside on right to the current operation if
+   * the graph is imagined to be in horizontal orientation.
+   * @param incomingOperations all the operations which are processed so far
+   * @param result a {@link Map} of {@link EndPointField} to {@link Set} of {@link EndPointField} which represents all
+   * the fields which have incoming connection the key field
+   */
+  private void computeIncoming(EndPointField destination, Operation currentOperation,
+                               Operation previousOperation, Set<Operation> incomingOperations,
+                               Map<EndPointField, Set<EndPointField>> result) {
+    if (!incomingOperations.add(currentOperation)) {
       return;
     }
 
+    // TODO (Rohit): This method is doing multiple unrelated things. First this part find outs all the incoming
+    // connections for computeIncomingSummary and it also finds the incoming operations for the field.
     if (currentOperation.getType() == OperationType.READ) {
       // if current operation is of type READ, previous operation must be of type TRANSFORM or WRITE
       // get only the input fields from the previous operations for which the origin is current READ operation
-      Set<InputField> inputFields = new HashSet<>();
-      if (OperationType.WRITE == previousOperation.getType()) {
-        WriteOperation previousWrite = (WriteOperation) previousOperation;
-        inputFields = new HashSet<>(previousWrite.getInputs());
-      } else if (OperationType.TRANSFORM == previousOperation.getType()) {
-        TransformOperation previousTransform = (TransformOperation) previousOperation;
-        inputFields = new HashSet<>(previousTransform.getInputs());
-      }
 
+      Set<InputField> inputFields = getInputFields(previousOperation);
```

Comment:
In this case, I think it's better not to have this additional method. It splits up logic that is more clear together. For example, when looking at the getInputFields method, it was not clear to me why OperationType.READ was invalid, but here it is clear, as we don't expect to have a READ connected to another READ.

Reply:
If anything, I would have a method to handle when the current operation is a read and a method to handle when it is a transform. Everything in the if clause could move into that method.